### PR TITLE
Improved error messages

### LINKED ISSUE #86: Improve error messages
# Overview

@Mikanebu 

Also, when I got an error message, it was not clear:
  * `Wrong type for header: Coordinates and value: -14.27667`
  Suggestion: `Coordinates field cannot accept value -14.27667. Geopoint type should be in the pattern "lon, lat".`
  * `Row X is completely blank
   {
    "error-count": 1,
    "headers": [
        "Country Name",
        "Country Code",
        "Year",
        "Value"
    ],
    "time": 0.406,
    "row-count": 1000,
    "source": "gdp/data/gdp.csv",
    "valid": false
    }`
  This error message has unnecassary information and not clear.
  Suggestion: `Row X is completely blank. According to the spec it cannot be accept empty rows. 
 Path to source`
  * `Row 7 is duplicated to row(s) 6`
  Same here   
  * `When schema type is not type of *array*`.
  Too much information and not clear.
  Suggestion: `Failed validating 'type' in schema['properties']['sources']. Sources property type must be type of array.`
  * `AttributeError: 'list' object has no attribute 'split'`
  Wrong geotype error which is not clear.
  Probably it is kind of bug

---

@roll

I agree on some but others are clearly from `goodtables` not from `tableschema-js`.

---

@Stephen-Gates 

Just wondering if the errors messages presented are using https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json?  

The messages above are similar to but not the same as the spec:

- `Row X is completely blank` is [L77](https://github.com/frictionlessdata/data-quality-spec/blob/cea0c664012d840d2d8d88ff6bce3f8b952476d7/spec.json#L77) but the [description](https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json#L78) doesn't seem to match
- `Row 7 is duplicate to row(s) 6` is [L85](https://github.com/frictionlessdata/data-quality-spec/blob/cea0c664012d840d2d8d88ff6bce3f8b952476d7/spec.json#L85) but the [description](https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json#L86) doesn't seem to match

These don't match the spec at all:
- `Wrong type for header: Coordinates and value: -14.27667`
- `When schema type is not type of *array*`
- `AttributeError: 'list' object has no attribute 'split'`

The [Data Quality Spec](https://github.com/frictionlessdata/data-quality-spec) seems to have a good set of messages and I imagine foreign key errors messages will be added soon. Should this library use the Data Quality Spec?

---
@Stephen-Gates on FK

message: The value {value} in row {row_number} and column {column_name_A} does not have a matching value in column {column_name_B} in table {table_name_B}

description: A matching value should exist for a valid foreign key relationship. How it could be resolved:

If this value is not correct, update the value.
If value is correct, then add it to column {column_name_B} in table {table_name_B}
If this error should be ignored, then disable the foreign key check for column {column_name_A}
Not sure if the last dot point is needed - just conforming with format of other messages.

### PR DESCRIPTION
**Pull request by roll**
- fixes #86

### CODE REVIEW

**Comment by roll** [comment]
@Stephen-Gates
Could you please take a look? I've improved some error messages (syncing some with Data Quality Spec). But still it's not exactly the same in high-level because various reasons like:
- fields-headers mismatch is just on error though DQS uses `extra/missing-header`
- not enough information for `field.castRow` to say on which row number is error
- unlike `goodtables` we handle constraints here together not one-by-one (still solvable)
- foreign key message requires some internal refactoring to requested message data
- etc

So here is a question how far we should improve it. Is it a real blocker for `data-curator`. Problem here that error messages here is a exceptional situation messages. It's not intended to be user-facing in general (e.g. app on top should catch it). Unlike to `goodtables` where all error messages are intended to be user-facing.

So looking forward to your comments. We still have some reserves to improve some messages more (investing more time). Just looking forward to hear what's blockers/priorities. We could handle it in this PR or open individual issue for every new requested improvement.

cc @pwalsh

**Comment by roll** [comment]
List of all data related exceptions:

- [field.js] 
- `The value "${value}" in column "${this.name}" is not type "${this.type}" and format "${this.format}"`
- `Field "${this.name}" has constraint "${name}" which is not satisfied for value "{value}"`
- [schema.js]
- `Row length ${row.length} doesn't match fields count ${this.fields.length}`
- [table.js]
- `Table headers don\'t match schema field names`
- `Row ${rowNumber} has unique constraint violation in column "${cache.name}"`
- `Foreign key "${foreignKey.fields}" violation in row ${rowNumber}`

**Comment by Stephen-Gates** [comment]
@roll I will have a good look over the weekend.

**Comment by Stephen-Gates** [comment]
@roll 

Ideally the row and column for each error would be reported. This would enable [Data Curator](https://github.com/ODIQueensland/data-curator#planned-features) to:

- highlight the cells where errors are found
- support a workflow where users validate a column at a time

If this is not possible, then returning the row in error would enable the row to be highlighted.

In terms of priority:

1. Type and Format 
2. Constraint 
3. Foreign Key
4. Structure

I'll add a feature request and comment separately on error message wording.

**Comment by roll** [comment]
@Stephen-Gates
That's what I was thinking for. Error messages may be are not the best thing to rely on. But the lib could provide an API for error context when possible e.g:
- error.row
- error.column
- error.code
- etc
Than it could be re-used in the higher-level app. I think it could be more reliable way to achieve the same goal.

**Comment by Stephen-Gates** [comment]
Sounds like a good approach @roll

**Comment by Stephen-Gates** [comment]
@roll sorry for the delay. Assuming no other variables are available for use in the error messages, then...

### Type or Format Error

`The value "${value}" in column "${this.name}" is not type "${this.type}" and format "${this.format}"` appears to be as close as you can get to the [Type or Format DQS error message in DQS](https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json#L138). 

### Constraint Error

To better match the format of most [constraint error messages in DQS](https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json#L159), I suggest 
`Field "${this.name}" has constraint "${name}" which is not satisfied for value "{value}"` 
is rewritten to,
`The value "{value}" does not conform to the "${name}" constraint for field "${this.name}"`

### Extra / Missing Value 
There's no equivalent message in DQS. The closest is the [extra/missing value error messages in DQS](https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json#L88) that compares the row length to the header row. This error appears to be comparing the row length to the number of fields in the schema. 

 I suggest
`Row length ${row.length} doesn't match fields count ${this.fields.length}` 
is rewritten to, 
`The row with ${row.length} columns does not match the ${this.fields.length} fields in the schema`

Perhaps `columns` should be `values`?

### Non-matching Header
To better match the format of most [non-matching header error message in DQS](https://github.com/frictionlessdata/data-quality-spec/blob/master/spec.json#L114), I suggest 
`Table headers don\'t match schema field names` 
is rewritten to, 
`The column header names do not match the field names in the schema` 

### Unique constraint

`Row ${rowNumber} has unique constraint violation in column "${cache.name}"`

edit that should be applied to DQS and here,

`Row ${rowNumber} has a unique constraint violation in column "${cache.name}"`

### Foreign key
Not implemented in DQS (although there is [a suggestion](https://github.com/frictionlessdata/data-quality-spec/issues/15#issuecomment-306777654)). I don't think I can improve 
`Foreign key "${foreignKey.fields}" violation in row ${rowNumber}` 

I did consider:
- `The foreign key "${foreignKey.fields}" in row ${rowNumber} has a foreign key violation`
- `The foreign key "${foreignKey.fields}" in row ${rowNumber} does not have a matching primary key value`

**Comment by roll** [comment]
@Stephen-Gates
Thanks. It's just great help.

**Comment by roll** [comment]
@Stephen-Gates 
Please take a look. I've:
- updated error messages based on your feedback
- added `error.rowNumber/columnNumber` properties (if available)

**Comment by Stephen-Gates** [comment]
@roll will review this week. thanks.

File: src/schema.js [hunk 1]
```diff
@@ -138,52 +139,31 @@ class Schema {
...
+        error.columnNumber = index + 1
+        if (failFast) throw error
+        errors.push(error)
       }
     }
 
-    if (errors.length > 0) {
+    // Raise errors
+    if (errors.length) {
       const message = `There are ${errors.length} cast errors (see 'error.errors')`
```
Comment (Stephen-Gates) [inline src/schema.js]: I wonder if this is better worded, `There are ${errors.length} type and format mismatch errors (see 'error.errors')`

**Comment by roll** [comment]
@Stephen-Gates 
Thanks! I've fixes the last one.

**Pull request merged**